core/services/health/health.go

```diff
@@ -0,0 +1,212 @@
+		// Register a service for health checks.
+		Register(name string, service Checkable) error
+		// Unregister a service.
+		Unregister(name string) error
+		// IsReady returns the current readiness of the system.
+		// A system is considered ready if all checks are passing (no errors)
+		IsReady() (ready bool, errors map[string]error)
+		// IsHealthy returns the current health of the system.
+		// A system is considered healthy if all checks are passing (no errors)
+		IsHealthy() (healthy bool, errors map[string]error)
+
+		Start() error
+		Close() error
+	}
+
+	checker struct {
+		srvMutex   sync.RWMutex
+		services   map[string]Checkable
+		stateMutex sync.RWMutex
+		state      map[string]State
+
+		stop chan struct{}
+		done chan struct{}
+
+		utils.StartStopOnce
+	}
+
+	State struct {
+		ready   error
+		healthy error
+	}
+
+	Status string
+)
+
+var _ Checker = (*checker)(nil)
+
+const (
+	StatusPassing Status = "passing"
+	StatusFailing Status = "failing"
+
+	interval = 15 * time.Second
+)
+
+var (
+	metrics = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "health",
+		Help: "Health status by service",
+	},
+		[]string{"service_id"},
+	)
+)
+
+func NewChecker() Checker {
+	c := &checker{
+		services: make(map[string]Checkable, 10),
+		state:    make(map[string]State, 10),
+		stop:     make(chan struct{}),
+		done:     make(chan struct{}),
+	}
+
+	return c
+}
+
+func (c *checker) Start() error {
+	return c.StartOnce("HealthCheck", func() error {
+		// update immediately
+		c.update()
+
+		go c.run()
+
+		return nil
+	})
+}
+
+func (c *checker) Close() error {
+	return c.StopOnce("HealthCheck", func() error {
+		close(c.stop)
+		<-c.done
+		return nil
+	})
+}
+
+func (c *checker) run() {
+	defer close(c.done)
+
+	ticker := time.NewTicker(interval)
+
+	for {
+		select {
+		case <-ticker.C:
+			c.update()
+		case <-c.stop:
+			return
+		}
+	}
+
+}
+
+func (c *checker) update() {
+	state := make(map[string]State, len(c.services))
+
+	c.srvMutex.RLock()
+
+	for name, s := range c.services {
+		ready := s.Ready()
+		healthy := s.Healthy()
+
+		state[name] = State{ready, healthy}
+	}
+
+	c.srvMutex.RUnlock()
+
+	// we use a separate lock to avoid holding the lock over state while talking
+	// to services
+	c.stateMutex.Lock()
+	defer c.stateMutex.Unlock()
+
+	for name, state := range state {
+		c.state[name] = state
+
+		value := 0
+		if state.healthy == nil {
+			value = 1
+		}
+
+		// report metrics to prometheus
+		metrics.WithLabelValues(name).Set(float64(value))
+	}
+}
+
+func (c *checker) Register(name string, service Checkable) error {
+	if service == nil || name == "" {
+		return errors.Errorf("misconfigured check %v", service)
+	}
+
+	c.srvMutex.Lock()
+	defer c.srvMutex.Unlock()
+	c.services[name] = service
+	return nil
+}
+
+func (c *checker) Unregister(name string) error {
+	if name == "" {
+		return errors.Errorf("name cannot be empty")
+	}
+
+	c.srvMutex.Lock()
+	defer c.srvMutex.Unlock()
+	c.services[name] = nil
```

Comment:
oh i think it should be `delete(c.service, name)` - otherwise the key will remain in the map